tests/test_torch.py

```diff
@@ -38,116 +35,129 @@
 
-        with self.assertRaisesRegex(QuantumFunctionError, 'can only be measured once'):
-            qf(par)
+        with pytest.raises(QuantumFunctionError, match='can only be measured once'):
+            qf(torch.tensor(0.5))
+
+    def test_qnode_fails_on_qfunc_with_too_many_wires(self, qubit_device_2_wires):
+        """The device must have sufficient wires for the qfunc"""
 
-        # device must have enough wires for the qfunc
-        @qml.qnode(self.dev2, interface='torch')
+        @qml.qnode(qubit_device_2_wires, interface='torch')
         def qf(x):
             qml.RX(x, wires=[0])
             qml.CNOT(wires=[0, 2])
             return qml.expval(qml.PauliZ(0))
 
-        with self.assertRaisesRegex(QuantumFunctionError, 'applied to invalid wire'):
-            qf(par)
+        with pytest.raises(QuantumFunctionError, match='applied to invalid wire'):
+            qf(torch.tensor(0.5))
 
-        # CV and discrete ops must not be mixed
-        @qml.qnode(self.dev1, interface='torch')
+    def test_qnode_fails_on_combination_of_cv_and_qbit_ops(self, qubit_device_1_wire):
+        """CV and discrete operations must not be mixed"""
+        
+        @qml.qnode(qubit_device_1_wire, interface='torch')
         def qf(x):
             qml.RX(x, wires=[0])
             qml.Displacement(0.5, 0, wires=[0])
             return qml.expval(qml.PauliZ(0))
 
-        with self.assertRaisesRegex(QuantumFunctionError, 'Continuous and discrete'):
-            qf(par)
+        with pytest.raises(QuantumFunctionError, match='Continuous and discrete'):
+            qf(torch.tensor(0.5))
 
-        # default plugin cannot execute CV operations, neither gates...
-        @qml.qnode(self.dev1, interface='torch')
+    def test_qnode_fails_for_cv_ops_on_qubit_device(self, qubit_device_1_wire):
+        """A qubit device can not execute CV operations"""
+
+        @qml.qnode(qubit_device_1_wire, interface='torch')
         def qf(x):
             qml.Displacement(0.5, 0, wires=[0])
             return qml.expval(qml.X(0))
 
-        with self.assertRaisesRegex(DeviceError, 'Gate [a-zA-Z]+ not supported on device'):
-            qf(par)
+        with pytest.raises(DeviceError, match='Gate [a-zA-Z]+ not supported on device'):
+            qf(torch.tensor(0.5))
+
+    def test_qnode_fails_for_cv_observables_on_qubit_device(self, qubit_device_1_wire):
+        """A qubit device can not measure CV observables"""
 
-        # ...nor observables
-        @qml.qnode(self.dev1, interface='torch')
+        @qml.qnode(qubit_device_1_wire, interface='torch')
         def qf(x):
             return qml.expval(qml.X(0))
 
-        with self.assertRaisesRegex(DeviceError, 'Observable [a-zA-Z]+ not supported on device'):
-            qf(par)
+        with pytest.raises(DeviceError, match='Observable [a-zA-Z]+ not supported on device'):
+            qf(torch.tensor(0.5))
```

Comment:
This is fantastic that you managed to separate them like this. A lot of work, but will help make the library a lot more maintainble

Reply:
lol who wrote them like this originally? :laughing:

Reply:
I mean it was not an unreasonable amount of work :+1: